python/tutorials/11-programmatic-dependent-launch.py

```diff
@@ -0,0 +1,240 @@
+The kernels support both FP16 and FP8 data types but the FP8 implementation is only available on CUDA devices with compute capability >= 9.0.
+
+.. code-block:: bash
+    python 11-programmatic-dependent-launch.py
+"""
+
+import torch
+import triton
+import triton.language as tl
+
+from typing import Optional
+
+
+def is_cuda():
+    return triton.runtime.driver.active.get_current_target().backend == "cuda"
+
+
+def supports_tma():
+    return is_cuda() and torch.cuda.get_device_capability()[0] >= 9
+
+
+def _matmul_launch_metadata(grid, kernel, args):
+    ret = {}
+    M, N, K = args["M"], args["N"], args["K"]
+    ret["name"] = f"{kernel.name} [M={M}, N={N}, K={K}]"
+    if "c_ptr" in args:
+        bytes_per_elem = args["c_ptr"].element_size()
+    elif "c_desc_ptr" in args:
+        bytes_per_elem = 2
+    else:
+        bytes_per_elem = 1 if args["FP8_OUTPUT"] else 2
+    ret[f"flops{bytes_per_elem * 8}"] = 2. * M * N * K
+    ret["bytes"] = bytes_per_elem * (M * K + N * K + M * N)
+    return ret
+
+
+def matmul_get_configs():
+    return [
+        triton.Config({'BLOCK_SIZE_M': BM, 'BLOCK_SIZE_N': BN, "BLOCK_SIZE_K" : BK, "USE_TMA" : USE_TMA}, num_stages=s, num_warps=w) \
+        for BM in [128] \
+        for BN in [128, 256] \
+        for BK in [64, 128] \
+        for USE_TMA in [True, False] \
+        for s in ([3, 4]) \
+        for w in [4, 8] \
+    ]
+
+
+@triton.autotune(
+    configs=matmul_get_configs(),
+    key=["M", "N", "K"],
+)
+@triton.jit(launch_metadata=_matmul_launch_metadata)
+def matmul_kernel(a_ptr, b_ptr, c_ptr,  #
+                  M, N, K,  #
+                  BLOCK_SIZE_M: tl.constexpr,  #
+                  BLOCK_SIZE_N: tl.constexpr,  #
+                  BLOCK_SIZE_K: tl.constexpr,  #
+                  USE_TMA: tl.constexpr,  #
+                  USE_GDC: tl.constexpr):  #
+    # This kernel can use TMA or traditional tl.load/tl.store
+    a_desc = None
+    b_desc = None
+    c_desc = None
+    a_ptrs = None
+    b_ptrs = None
+    c_ptrs = None
+
+    dtype = c_ptr.dtype.element_ty
+    start_pid = tl.program_id(axis=0)
+    num_pid_m = tl.cdiv(M, BLOCK_SIZE_M)
+    k_tiles = tl.cdiv(K, BLOCK_SIZE_K)
+
+    pid_m = start_pid % num_pid_m
+    pid_n = start_pid // num_pid_m
+
+    if USE_TMA:
+        offs_am = pid_m * BLOCK_SIZE_M
+        offs_bn = pid_n * BLOCK_SIZE_N
+    else:
+        offs_am = pid_m * BLOCK_SIZE_M + tl.arange(0, BLOCK_SIZE_M)
+        offs_bn = pid_n * BLOCK_SIZE_N + tl.arange(0, BLOCK_SIZE_N)
+        offs_am = tl.where(offs_am < M, offs_am, 0)
+        offs_bn = tl.where(offs_bn < N, offs_bn, 0)
+        offs_am = tl.max_contiguous(tl.multiple_of(offs_am, BLOCK_SIZE_M), BLOCK_SIZE_M)
+        offs_bn = tl.max_contiguous(tl.multiple_of(offs_bn, BLOCK_SIZE_N), BLOCK_SIZE_N)
+        offs_k = tl.arange(0, BLOCK_SIZE_K)
+        a_ptrs = a_ptr + (offs_am[:, None] * K + offs_k[None, :])
+        # B is N x K
+        b_ptrs = b_ptr + (offs_k[None, :] + offs_bn[:, None] * K)
```

Comment:
can we simplify this and only support TMA or not TMA. Tutorials are meant to be simple and showcase a single thing. I would try to make it minimal. If you just want to make an example making a minimal test is better.

Reply:
I simplified the tutorial to just reuse the vector add tutorial with and without PDL.